File: franchises.go
package main

// franchises.go reads a checkout of a franchises repository into the three
// franchise tables, the way movies.go reads a title folder. The checkout
// holds one directory per franchise, each with a franchise.yaml and the
// franchise's art beside it under Kodi's names. The walk reads no volume
// and no other library, because a member is a provider alias the member's
// own library writes. A file the schema refuses is counted unidentified and
// skipped, and the files beside it still write their rows.

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
)

// scopeFranchise is the word that leads every franchise id, beside the
// scopes in rows.go.
const scopeFranchise = "franchise"

// franchiseFileName is the one file the scanner reads out of a franchise
// directory. The AGENTS.md beside it is for the next author, and the art
// is for the enrichers.
const franchiseFileName = "franchise.yaml"

// franchiseBody is what the page draws around the wall, as the franchises
// row's body. The names are the file's own, so a reader of the row and a
// reader of the file read one vocabulary.
type franchiseBody struct {
	Universe string             `json:"universe,omitempty"`
	Calendar *franchiseCalendar `json:"calendar,omitempty"`
	Eras     []franchiseEra     `json:"eras,omitempty"`
	Sources  []string           `json:"sources,omitempty"`
}

// franchiseRow is one row of the franchises item table: the item header
// every kind carries, with the franchise's clock in the body. Released is
// empty, because the scanner reads no member's date.
type franchiseRow struct {
	Id       string
	Library  string
	Kind     string
	Path     string
	Title    string
	SortKey  string
	Slug     string
	Released string
	Added    int64
	Art      string
	Arts     []string
	Duration int64
	Body     franchiseBody
}

// franchiseMemberRow is one entry of one franchise's order, at its position
// in story order. Alias is the member as the provider alias its own library
// writes, and it is the whole join to the catalog. Universes is a JSON
// list, empty where the entry is in the home universe alone. ReleaseYear is
// the file's own, and 0 where the file gives none.
type franchiseMemberRow struct {
	Library   string
	Franchise string
	Position  int
	Kind      string
	Alias     string
	Title     string
	Released  string
	// ReleaseYear is the year of that date, its first four characters, so
	// the wall reads one integer for its label and never parses a string.
	ReleaseYear int
	Timed       int
	TimeFrom    float64
	TimeTo      float64
	Universes   string
}

// franchiseRunRow is one season, or one episode, of one series run.
// Episode 0 is the whole season, and a run with no rows is the whole show.
type franchiseRunRow struct {
	Library   string
	Franchise string
	Position  int
	Season    int
	Episode   int
}

// franchiseID derives the id of one franchise from the name of its
// directory. The file carries no provider id, so the directory name is the
// identity, and a renamed directory is a new row.
func franchiseID(directory string) string {
	return scopeFranchise + ":name:" + slug(directory, 0)
}

// walkFranchises reads a whole checkout into one walkResult, one directory
// at a time. The directories are read in name order, so two walks of one
// commit write the same rows in the same order. A directory with no
// franchise.yaml is not a franchise, which is how the repository's own
// .git directory is passed over. The checkout holds the franchise.yaml
// files, and the claim holds the art the scan downloaded from the links
// each one carries; a franchise's directory carries the same name in both.
func walkFranchises(checkout, artRoot, library string) *walkResult {
	result := &walkResult{}
	names, err := franchiseDirectories(checkout)
	if err != nil {
		result.noteReadError(err)
		return result
	}
	for _, name := range names {
		scanFranchiseDirectory(checkout, artRoot, name, library, result)
	}
	return result
}

// franchiseDirectories are the directories of one checkout, in name order, so
// two walks of one commit read them the same way. A dot name is passed over,
// which is how the repository's own .git directory is left out.
func franchiseDirectories(checkout string) ([]string, error) {
	entries, err := os.ReadDir(checkout)
	if err != nil {
		return nil, err
	}
	names := []string{}
	for _, entry := range entries {
		if entry.IsDir() && !skipName(entry.Name()) {
			names = append(names, entry.Name())
		}
	}
	sort.Strings(names)
	return names, nil
}

// readFranchiseFile reads and validates one directory's franchise.yaml. A
// directory that holds no franchise.yaml is not a franchise, and it returns no
// file and no error.
func readFranchiseFile(checkout, name string) (*franchiseFile, error) {
	data, err := os.ReadFile(filepath.Join(checkout, name, franchiseFileName))
	if errors.Is(err, fs.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("%w: %v", errFranchiseUnreadable, err)
	}
	return parseFranchiseFile(data)
}

// errFranchiseUnreadable tells a file the scanner could not read from a file
// the schema refuses. The first marks the pass incomplete, and the second is
// reported and skipped.
var errFranchiseUnreadable = errors.New("the scanner could not read the file")

// scanFranchiseDirectory reads one franchise directory into its rows. A
// file the schema refuses leaves the directory counted unidentified and
// named, the same reporting path a folder no sidecar identifies takes. A
// file the scanner cannot read marks the pass incomplete, so a checkout it
// could not read never sweeps the rows the catalog holds.
func scanFranchiseDirectory(checkout, artRoot, name, library string, result *walkResult) {
	file, err := readFranchiseFile(checkout, name)
	if err != nil {
		// A file the scanner could read and the schema refuses is a fault
		// in the file; a file it could not read is a fault in the checkout.
		if errors.Is(err, errFranchiseUnreadable) {
			result.noteReadError(err)
			return
		}
		result.unidentified++
		result.unidentifiedNames = append(result.unidentifiedNames, name)
		return
	}
	if file == nil {
		return
	}

	id := franchiseID(name)
	// The art is read off the claim, the way every other kind reads it, so the
	// columns hold paths under the library root and never links.
	primaryArt, allArt, err := franchiseArt(artRoot, name)
	result.noteReadError(err)

	result.franchises = append(result.franchises, franchiseRow{
		Id:      id,
		Library: library,
		Kind:    libraryKindFranchises,
		Path:    name,
		Title:   file.Name,
		SortKey: sortKey(file.Name),
		Slug:    slug(file.Name, 0),
		Art:     primaryArt,
		Arts:    allArt,
		Body: franchiseBody{
			Universe: file.Universe,
			Calendar: file.Calendar,
			Eras:     file.Eras,
			Sources:  file.Sources,
		},
	})
	appendFranchiseOrder(library, id, file, result)
	result.titles++
}

// appendFranchiseOrder walks the order and writes one member row per
// entry, from position 1. A series with seasons is one member row and one
// run row per season or episode, so The Clone Wars is one slot on the wall
// and not thirty.
func appendFranchiseOrder(library, franchise string, file *franchiseFile, result *walkResult) {
	for index, entry := range file.Order {
		position := index + 1
		result.franchiseMembers = append(result.franchiseMembers,
			franchiseMemberRow{
				Library:     library,
				Franchise:   franchise,
				Position:    position,
				Kind:        entry.kind(),
				Alias:       entry.kind() + ":" + entry.reference(),
				Title:       entry.Title,
				Released:    entry.Released,
				ReleaseYear: entry.releaseYear(),
				Timed:       timedMark(entry.Time),
				TimeFrom:    spanEnd(entry.Time, func(t franchiseTime) *float64 { return t.From }),
				TimeTo:      spanEnd(entry.Time, func(t franchiseTime) *float64 { return t.To }),
				Universes:   universesValue(entry.Universes),
			})
		for _, season := range entry.Seasons {
			result.franchiseRuns = append(result.franchiseRuns,
				franchiseRunRows(library, franchise, position, season)...)
		}
	}
}

// franchiseRunRows are the run rows of one season: the whole season, or
// one row per episode the file names. A range expands here, because the
// file expands with no catalog, and a held-episode count is then one join.
func franchiseRunRows(library, franchise string, position int, season franchiseSeason) []franchiseRunRow {
	row := franchiseRunRow{Library: library, Franchise: franchise, Position: position, Season: *season.Season}
	if len(season.Episodes) == 0 {
		return []franchiseRunRow{row}
	}
	rows := []franchiseRunRow{}
	for _, code := range season.Episodes {
		// The file validated before the walk read it, so every code
		// expands and the error here cannot happen.
		episodes, _ := expandEpisodeCode(code)
		for _, episode := range episodes {
			rows = append(rows, franchiseRunRow{
				Library: library, Franchise: franchise, Position: position,
				Season: episode.Season, Episode: episode.Episode,
			})
		}
	}
	return rows
}

// timedMark is 1 where the entry carries a time, and 0 where it carries
// none.
func timedMark(span *franchiseTime) int {
	if span == nil {
		return 0
	}
	return 1
}

// spanEnd is one end of an entry's span, and 0 where the entry has no
// time. The file validated before the walk read it, so a span the walk
// reads carries both ends.
func spanEnd(span *franchiseTime, end func(franchiseTime) *float64) float64 {
	if span == nil {
		return 0
	}
	if value := end(*span); value != nil {
		return *value
	}
	return 0
}

// universesValue is the entry's universes as the JSON list the column
// holds. An entry that names none is in the home universe alone, which is
// the empty list.
func universesValue(universes []string) string {
	if len(universes) == 0 {
		return "[]"
	}
	payload, _ := json.Marshal(universes)
	return string(payload)
}

File: media-browser/src/catalog/sidecar/franchises.rs
// The two reads a media browser makes of a franchise, as catalog.sql writes
// them beside the tables. The strip read is the INNER JOIN: the whole order
// with only the members some library holds. The page read is the LEFT JOIN:
// every entry, held or not, with the episodes the catalog holds for a series
// run. Both resolve a member through the aliases table across every library of
// the namespace, and the first library by name wins where two hold one member.

use rusqlite::{Connection, Row};
use serde_json::Value;

use super::collect;
use super::item;
use crate::catalog::FranchiseEntry;
use crate::catalog::franchise::{Calendar, Entry, Era, Franchise, Held, Membership};

// The two item tables as one list of members, which both reads join
// through. The kind column names the table a row came from, so a press
// on a slot opens the page of its own kind.
const ITEMS: &str = "SELECT library, id, title, art, arts, released, slug, body, duration, \
                            'movies' AS kind \
                     FROM movies \
                     UNION ALL \
                     SELECT library, id, title, art, arts, released, slug, body, duration, \
                            'series' AS kind \
                     FROM series";

// The member columns both reads answer with, in the order [`entry`]
// takes them. The last of them is the held item, which is NULL for
// every gap of the page read.
const COLUMNS: &str = "m.position, m.kind, m.alias, m.title, m.release_year, m.timed, \
                       m.time_from, m.time_to, m.universes, MIN(i.library), i.id, i.title, \
                       i.art, i.arts, i.released, i.slug, i.kind, m.released, i.duration";

/// Every franchise row of every library, in sort order. The read is one scan
/// of the franchises table, which holds one row per franchise of the namespace
/// and is the smallest table of the catalog, so it needs no index of its own.
pub fn all(connection: &Connection) -> rusqlite::Result<Vec<FranchiseEntry>> {
    // Where the franchises row carries no art of its own, the page falls
    // back to the poster of the first member some library holds. The
    // member comes off the same alias join the other two reads make, in
    // story order, and the first one with a poster wins. The poster
    // resolves against that member's own library, which is the second
    // subquery.
    let member = |column: &str| {
        format!(
            "(SELECT {column} FROM franchise_members m \
              JOIN aliases a ON a.alias = m.alias \
              JOIN items i ON i.library = a.library AND i.id = a.item \
              WHERE m.library = f.library AND m.franchise = f.id AND i.art != '' \
              ORDER BY m.position, i.library LIMIT 1)"
        )
    };
    let sql = format!(
        "WITH items AS ({ITEMS}) \
         SELECT f.library, f.id, f.title, f.art, f.slug, {art}, {holder} \
         FROM franchises f ORDER BY f.sort_key, f.library, f.id",
        art = member("i.art"),
        holder = member("i.library"),
    );
    collect(connection, &sql, &[], |row| {
        let own: String = item::text(row, 3)?;
        let (art, art_library) = match own.is_empty() {
            true => (item::text(row, 5)?, item::text(row, 6)?),
            false => (own, row.get(0)?),
        };
        Ok(FranchiseEntry {
            library: row.get(0)?,
            id: row.get(1)?,
            title: row.get(2)?,
            art,
            art_library,
            slug: item::text(row, 4)?,
        })
    })
}

/// Every franchise this item belongs to, with its held members in position
/// order. The franchises come from the item's own aliases, and the join to the
/// item tables keeps the members some library holds. The franchise's title
/// comes off the franchises row, because the strip's heading is that title.
/// The read also counts the entries of the order by kind, held or not, on
/// the primary key of `franchise_members`, so the heading says the scope of
/// the order and not only the members the namespace holds.
pub fn strips(
    connection: &Connection,
    library: &str,
    id: &str,
) -> rusqlite::Result<Vec<Membership>> {
    let sql = format!(
        "WITH mine AS (\
           SELECT alias FROM aliases WHERE library = ?1 AND item = ?2\
         ), found AS (\
           SELECT DISTINCT library, franchise FROM franchise_members \
           WHERE alias IN (SELECT alias FROM mine)\
         ), items AS ({ITEMS}) \
         SELECT m.library, m.franchise, \
           (SELECT title FROM franchises named \
            WHERE named.library = m.library AND named.id = m.franchise), \
           (SELECT count(*) FROM franchise_members every \
            WHERE every.library = m.library AND every.franchise = m.franchise \
              AND every.kind = 'movie'), \
           (SELECT count(*) FROM franchise_members every \
            WHERE every.library = m.library AND every.franchise = m.franchise \
              AND every.kind = 'series'), \
           {COLUMNS} \
         FROM franchise_members m \
         JOIN found f ON f.library = m.library AND f.franchise = m.franchise \
         JOIN aliases a ON a.alias = m.alias \
         JOIN items i ON i.library = a.library AND i.id = a.item \
         GROUP BY m.library, m.franchise, m.position \
         ORDER BY m.library, m.franchise, m.position"
    );
    let members = collect(connection, &sql, &[&library, &id], |row| {
        Ok((
            Membership {
                library: row.get(0)?,
                id: row.get(1)?,
                title: item::text(row, 2)?,
                movies: row.get(3)?,
                series: row.get(4)?,
                members: Vec::new(),
            },
            entry(row, 5)?,
        ))
    })?;

    // The read answers one row per member, ordered by the franchise, so
    // the members fold into the franchise the row before them opened.
    let mut strips: Vec<Membership> = Vec::new();
    for (franchise, member) in members {
        match strips.last_mut() {
            Some(last) if last.library == franchise.library && last.id == franchise.id => {}
            _ => strips.push(franchise),
        }
        if let Some(last) = strips.last_mut() {
            last.members.push(member);
        }
    }
    Ok(strips)
}

/// One franchise as a list of one, and an empty list where that `Library`
/// holds no franchise under that id. The header comes off the franchises row,
/// and the body carries the universe, the calendar, and the eras as the file
/// wrote them.
pub fn franchise(
    connection: &Connection,
    library: &str,
    id: &str,
) -> rusqlite::Result<Vec<Franchise>> {
    let sql = "SELECT title, art, json_extract(body, '$.universe'), \
                      json_extract(body, '$.calendar'), json_extract(body, '$.eras') \
               FROM franchises WHERE library = ? AND id = ?";
    let mut found = collect(connection, sql, &[&library, &id], |row| {
        Ok(Franchise {
            library: library.to_string(),
            id: id.to_string(),
            title: row.get(0)?,
            art: item::text(row, 1)?,
            universe: item::text(row, 2)?,
            calendar: calendar(&item::text(row, 3)?),
            eras: eras(&item::text(row, 4)?),
            entries: Vec::new(),
        })
    })?;

    if let Some(page) = found.first_mut() {
        page.entries = entries(connection, library, id)?;
    }
    Ok(found)
}

// Every entry of one franchise in story order, held or not. The
// episodes column counts what the catalog holds for a series run: every
// episode of the show where the run names no season, and the episodes
// the runs name where it does. The read also takes the tagline and the
// plot out of the held item's body, because the wall's card draws them
// beside the art; the strip read carries neither.
fn entries(connection: &Connection, library: &str, id: &str) -> rusqlite::Result<Vec<Entry>> {
    let sql = format!(
        "WITH items AS ({ITEMS}) \
         SELECT {COLUMNS}, \
           (SELECT count(*) FROM episodes e \
            JOIN aliases sa ON sa.library = e.library AND sa.item = e.series \
            WHERE sa.alias = m.alias \
              AND (NOT EXISTS (SELECT 1 FROM franchise_runs r \
                               WHERE r.library = m.library AND r.franchise = m.franchise \
                                 AND r.position = m.position) \
                   OR EXISTS (SELECT 1 FROM franchise_runs r \
                              WHERE r.library = m.library AND r.franchise = m.franchise \
                                AND r.position = m.position AND r.season = e.season \
                                AND r.episode IN (0, e.episode)))) AS held_episodes, \
           json_extract(i.body, '$.tagline'), json_extract(i.body, '$.plot') \
         FROM franchise_members m \
         LEFT JOIN aliases a ON a.alias = m.alias \
         LEFT JOIN items i ON i.library = a.library AND i.id = a.item \
         WHERE m.library = ?1 AND m.franchise = ?2 \
         GROUP BY m.position \
         ORDER BY m.position"
    );
    collect(connection, &sql, &[&library, &id], |row| {
        let mut member = entry(row, 0)?;
        member.episodes = row.get(19)?;
        if let Some(held) = member.held.as_mut() {
            held.tagline = item::text(row, 20)?;
            held.plot = item::text(row, 21)?;
        }
        Ok(member)
    })
}

// One member row as an entry, from the columns [`COLUMNS`] names,
// starting at `at`. A held item's own library is the MIN over the
// libraries that hold it, and a NULL there is a gap.
fn entry(row: &Row<'_>, at: usize) -> rusqlite::Result<Entry> {
    let library: Option<String> = row.get(at + 9)?;
    let held = match library {
        Some(library) => Some(Held {
            library,
            id: item::text(row, at + 10)?,
            kind: item::text(row, at + 16)?,
            title: item::text(row, at + 11)?,
            art: item::text(row, at + 12)?,
            arts: item::strings(&item::text(row, at + 13)?),
            released: item::text(row, at + 14)?,
            slug: item::text(row, at + 15)?,
            tagline: String::new(),
            plot: String::new(),
            duration: row.get::<_, Option<i64>>(at + 18)?.unwrap_or_default(),
        }),
        None => None,
    };
    Ok(Entry {
        position: row.get(at)?,
        kind: row.get(at + 1)?,
        alias: row.get(at + 2)?,
        title: row.get(at + 3)?,
        released: item::text(row, at + 17)?,
        release_year: row.get(at + 4)?,
        timed: row.get::<_, i64>(at + 5)? != 0,
        from: row.get(at + 6)?,
        to: row.get(at + 7)?,
        universes: item::strings(&item::text(row, at + 8)?),
        held,
        episodes: 0,
    })
}

// The body's calendar, and nothing where the file names none. A
// calendar needs a unit, so a block without one is no calendar.
fn calendar(json: &str) -> Option<Calendar> {
    let Ok(Value::Object(block)) = serde_json::from_str::<Value>(json) else {
        return None;
    };
    let unit = word(&block, "unit");
    match unit.is_empty() {
        true => None,
        false => Some(Calendar {
            unit,
            zero: word(&block, "zero"),
            before: word(&block, "before"),
            after: word(&block, "after"),
        }),
    }
}

// The body's eras, in the order the file names them. An era with no
// name is left out, because a bar with no words on it says nothing.
fn eras(json: &str) -> Vec<Era> {
    let Ok(Value::Array(named)) = serde_json::from_str::<Value>(json) else {
        return Vec::new();
    };
    named
        .iter()
        .filter_map(|era| {
            let block = era.as_object()?;
            let name = word(block, "name");
            match name.is_empty() {
                true => None,
                false => Some(Era {
                    name,
                    from: era.get("from").and_then(Value::as_f64).unwrap_or_default(),
                    to: era.get("to").and_then(Value::as_f64).unwrap_or_default(),
                }),
            }
        })
        .collect()
}

fn word(block: &serde_json::Map<String, Value>, name: &str) -> String {
    block
        .get(name)
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_body_with_no_calendar_answers_none() {
        assert_eq!(calendar(""), None);
        assert_eq!(calendar("null"), None);
        assert_eq!(calendar("{}"), None);
        assert_eq!(calendar(r#"{"zero":"Yavin"}"#), None);
    }

    #[test]
    fn a_calendar_carries_its_unit_and_its_two_words() {
        assert_eq!(
            calendar(r#"{"unit":"years","zero":"Yavin","before":"BBY","after":"ABY"}"#),
            Some(Calendar {
                unit: "years".into(),
                zero: "Yavin".into(),
                before: "BBY".into(),
                after: "ABY".into(),
            })
        );
    }

    #[test]
    fn the_eras_come_in_the_order_the_file_names_them() {
        let eras = eras(r#"[{"name":"Late","from":-5,"to":5},{"name":"Early","from":-500}]"#);
        assert_eq!(
            eras,
            [
                Era {
                    name: "Late".into(),
                    from: -5.0,
                    to: 5.0
                },
                Era {
                    name: "Early".into(),
                    from: -500.0,
                    to: 0.0
                },
            ]
        );
    }

    #[test]
    fn an_era_with_no_name_and_a_body_with_no_eras_leave_nothing_behind() {
        assert!(eras("").is_empty());
        assert!(eras("null").is_empty());
        assert!(eras(r#"{"name":"Late"}"#).is_empty());
        assert!(eras(r#"[{"from":-5,"to":5},"Late"]"#).is_empty());
    }
}
